packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -175,26 +249,271 @@ impl HookContextImpl {
 /// </pre>
 ///
 /// [`Debug`]: core::fmt::Debug
+// TODO: ideally we would want to make `HookContextInner` private, as it is an implementation
+//  detail, but "attribute privacy" as outlined in https://github.com/rust-lang/rust/pull/61969
+//  is currently not implemented for repr(transparent).
 #[repr(transparent)]
-pub struct HookContext<'a, T> {
-    parent: &'a mut HookContextImpl,
-    _marker: PhantomData<T>,
+pub struct HookContext<T> {
+    inner: HookContextInner,
+    _marker: PhantomData<fn(&T)>,
 }
 
-impl<T> HookContext<'_, T> {
+impl<T> HookContext<T> {
+    pub(crate) fn new(alternate: bool) -> Self {
+        Self {
+            inner: HookContextInner::new(alternate),
+            _marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn append_snippet_string(&mut self, snippets: &mut Vec<String>) {
+        self.inner.append_snippet_string(snippets);
+    }
+
+    pub(crate) fn snippet_strings(&self) -> &[String] {
+        self.inner.snippet_strings()
+    }
+
     /// This snippet (which can include line breaks) will be appended to the
     /// main message.
     ///
     /// This is useful for dense information like backtraces, or span traces.
     ///
     /// [`alternate()`]: Self::alternate
     /// [`Debug`]: core::fmt::Debug
-    pub fn attach_snippet(&mut self, snippet: impl Into<String>) {
-        self.parent.snippets.push(snippet.into());
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # // we only test with Rust 1.65, which means that `render()` is unused on earlier version
+    /// # #![cfg_attr(not(rust_1_65), allow(dead_code, unused_variables, unused_imports))]
+    /// use std::io::ErrorKind;
+    ///
+    /// use error_stack::Report;
+    ///
+    /// struct Error {
+    ///     code: usize,
+    ///     reason: &'static str,
+    /// }
+    ///
+    /// Report::install_debug_hook::<Error>(|Error { code, reason }, ctx| {
+    ///     if ctx.alternate() {
+    ///         ctx.snippet(format!("Error {code}:\n  {reason}"));
+    ///     }
```

Comment:
Trying to highlight the actual function in the example
```suggestion
    /// Report::install_debug_hook::<Error>(|Error { code, reason }, ctx| {
    ///     if ctx.alternate() {
    ///         // Add a snippet to the output
    ///         ctx.snippet(format!("Error {code}:\n  {reason}"));
    ///     }
```